src/main/java/com/radixdlt/atommodel/message/MessageParticle.java

```diff
@@ -111,4 +111,28 @@ public String toString() {
 		return String.format("%s[(%s:%s)]",
 			getClass().getSimpleName(), String.valueOf(from), String.valueOf(to));
 	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (!(o instanceof MessageParticle)) {
+			return false;
+		}
+		MessageParticle that = (MessageParticle) o;
+		return nonce == that.nonce
+				&& Objects.equals(from, that.from)
+				&& Objects.equals(to, that.to)
+				&& Objects.equals(metaData, that.metaData)
+				&& Arrays.equals(bytes, that.bytes)
+				&& Objects.equals(getDestinations(), that.getDestinations());
```

Comment:
Would rather not have all these equals() and hashCode() be refactored as these just make it brittle requiring to check parent classes. For example, if parent class changes then all children classes have incorrect equals().

Reply:
Also, in the long term, particles will be treated less like POJOs and more like a data chunks in which case equals and hashCode should be dependent only on hash of data chunks.

Reply:
@talekhinezh Any suggestion how to make it better? It used to rely on inherited equals&hashCode that worked on a calculated hash, but now we got rid of the global Hasher instance so that's no longer possible.

Reply:
ah, good point, I guess we can live with this for now.

In this case, instead of comparing items in the base class (`getDestinations()`) better to use `super` for comparison:
```
        if (!super.equals(obj)) return false;
```

Reply:
Same goes for hashCode

Reply:
are you sure about that?
currently `Particle` is abstract and it doesn't have any equals impl (it's supposed to be implemented in all subclasses).

It doesn't sound right if I add a default impl there (say: `o instanceof Particle && destinations.equals(that.destinations))` and your line in subclasses: `if (!super.equals(obj)) return false;`

because then we lose equals symmetry, f.e.:
```
Particle p1 = new Particle { destinations = abc }
Particle p2 = new MessageParticle(destinations = abc, ..other fields.)

p1.equals(p2) // true (only checks `destinations`)
p2.equals(p1) // false (checks parent destinations (super.equals) - true, AND concrete subclass fields - false)
```

Reply:
Ah yes, you're right. Another reason why inheritance so bad :) Okay, let's go with your solution.